src/main/scala/latis/service/hapi/DataService.scala

```diff
@@ -25,33 +26,44 @@ class DataService[F[_]: Concurrent](
   val service: HttpRoutes[F] =
     HttpRoutes.of[F] {
       case GET -> Root / "data"
-          :? IdMatcher(_id)
+          :? DatasetMatcher(_dataset)
+          +& IdMatcher(_id)
+          +& StartTimeMatcher(_startTime)
+          +& StopTimeMatcher(_stopTime)
           +& MinTimeMatcher(_minTime)
           +& MaxTimeMatcher(_maxTime)
           +& ParamMatcher(_params)
           +& IncludeMatcher(_inc)
           +& FormatMatcher(_fmt) =>
-        val req: Either[Status, DataRequest] =
+        val req: Either[Status, DataRequest] = {
+          // For backwards compatibility, allow both
+          // "dataset" or "id",
+          // "start" or "time.min",
+          // "stop" or "time.max"
           for {
-            id      <- _id.asRight
-            minTime <- _minTime.leftMap(_ => Status.`1402`).toEither
-            maxTime <- _maxTime.leftMap(_ => Status.`1403`).toEither
-            _       <- Either.cond(minTime.isBefore(maxTime), (), Status.`1404`)
-            params  <- Either.cond(
+            dataset   <- _dataset.getOrElse(_id.getOrElse("")).asRight
+            startTime <- _startTime.getOrElse(_minTime.getOrElse(LocalDateTime.MAX.invalid))
+              .leftMap(_ => Status.`1402`).toEither
+            stopTime  <- _stopTime.getOrElse(_maxTime.getOrElse(LocalDateTime.MIN.invalid))
+              .leftMap(_ => Status.`1403`).toEither
```

Comment:
Try to apply what I wrote about above to these. It's a little trickier because you have two layers of types (`Option` and `ValidatedNel`), but you can do it cleanly without binding a temporary variable.

As a hint, I'll say that we don't care why a time failed to parse, so you can safely lift the `ValidatedNel` into a type that forgets the reason.